### ISSUE TITLE
Typehinting the return value from a singleton in ServiceProviders

### ISSUE THREAD

**Issue by gms8994** [nunomaduro/larastan#1597]
- Larastan Version: 2.3.4
- `--level` used: 6
- Pull request with failing test:

### Description

Given the `ContainerProvider` below, and the singleton defined in `AppServiceProvider`, how do I typehint that the singleton doesn't return an instance of `ContainerProvider` but instead returns a `\Psr\Container\ContainerInterface` so that I don't have to litter my general code with

```
/** @var \Psr\Container\ContainerInterface */
$container = app()->make(ContainerProvider::class);
```

### Laravel code where the issue was found

```php
<?php

class ContainerProvider
{
	/** @var \Psr\Container\ContainerInterface */
	private static $container;
	
	/**
	 * @return \Psr\Container\ContainerInterface
	 */
	public static function getContainer()
	{
		if (null === self::$container) {
			self::initializeContainer();
		}

		return self::$container;
	}

	private static function initializeContainer() : void {
		self::$container = new \Pimple\Psr11\Container(new \Pimple\Container());
	}
}

class AppServiceProvider extends ServiceProvider
{
	public function register() : void {
		$this->app->singleton(ContainerProvider::class, function ($app) {
			return ContainerProvider::getContainer();
		});
	}
}
```

**Comment by szepeviktor** [comment]
Maybe `$this->app->singleton(ContainerProvider::class, function ($app): ContainerInterface {`

**Comment by mad-briller** [comment]
if you `PHPStan\dumpType($container` what do you get ?

**Comment by mad-briller** [comment]
also you bind `ContainerProvider` but it actually returns a `ContainerInterface`, this in general is a bit misleading as you'd expect to get the type you ask for back when using classnames in `->make()`

**Comment by gms8994** [comment]
> Maybe `$this->app->singleton(ContainerProvider::class, function ($app): ContainerInterface {`

I did try that, and sadly it didn't make a difference.

> if you `PHPStan\dumpType($container)` without the `@var` what do you get ?

It shows `ContainerProvider`

> also you bind `ContainerProvider` but it actually returns a `ContainerInterface`, this in general is a bit misleading as you'd expect to get the type you ask for back when using classnames in `->make()`

Yeah; this is some legacy code that we're trying to clear out, but it's a long time coming.

**Comment by mad-briller** [comment]
can you try larastan `2.3.5` atleast and see if the issue persists? as of 2.3.x larastan should resolve items from the container itself to inform the return types:

https://github.com/nunomaduro/larastan/blob/master/src/ReturnTypes/AppMakeHelper.php#L32-L47

a fix was made in 2.3.5 to handle it better, so perhaps that may fix the issue ?

**Comment by gms8994** [comment]
2.3.5 does seem to work, thank you!
